### PR TITLE
Naming refactor and tests, part 2

### PR DESCRIPTION
**Pull request by mtrmac**
Following up on #1112, this:
- Adds similar tests to the other half of the `pullImage` heuristics. (As before, they document the existing practice and allow noticing changes; they _don’t_ claim that the existing practice is correct.)
- Explicitly codifies the “`docker-archive:` pulls copy all names” and “when using search registries, error messages point to the config file” behaviors at the time these features are directly used, instead of adding extraordinary redundant logic in the core pull loop.
- Stops passing around so many redundant parameters when constructing `CopyOptions` values.
- Avoids some redundant string formatting + parsing roundtrips.
- Finally, some notable payoff: `podman load` and `podman save` use well-typed `types.ImageReference` instead of constructing strings which may be heuristically misinterpreted.

Still only small steps to untangling `ImageName` and `imageParts` I’m afraid…

Behavior changes, aside from outright bug fixes, should be very slight (changed error message here and there), apart from `podman pull docker-archive:…`, which now treats the input as a `docker-archive:` transport reference, without involving the Docker-like reference parsers and heuristics (e.g. for input like `docker-archive:latest`).

Please see individual commit messages for details; pay particular attention to the RFC ones.

This adds a few `FIXME`s to the non-test code, I hope to clean that up soonish.

(Also, is there a convenient way to set up a `containers-storage:` store as a completely non-privileged user (not root, no user namespaces, works on macOS)? I’d love to get rid of `pullRefName` and test the real values.)

### CODE REVIEW

**Comment by mheon** [comment]
I'll do a more thorough review on monday, but initial pass looks good.

I'm a little discouraged that references have started to enter the API - the original intent was to try and insulate people from the large amount of potentially-confusing reference types as much as possible. But it doesn't seem like it can be avoided - attempting to do this with strings is like pulling teeth.

**Comment by mtrmac** [comment]
WIP notes to self:

- `c/image/storage/storageTransport` tests run as non-root, reuse that here and drop `refNamePair`
- Add `c/image/docker/archive.NewReference` to remove two more string parsing cases.

File: cmd/podman/pull.go [hunk 1]
```diff
@@ -109,8 +110,12 @@ func pullCmd(c *cli.Context) error {
 	}
 
 	// Possible for docker-archive to have multiple tags, so use NewFromLoad instead
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply cmd/podman/pull.go]: The comment is roughly right, except the `NewFromLoad` name was mismatched with real function name from the start; updated the comment in the commit which renames the function to now match.

File: cmd/podman/push.go [hunk 2]
```diff
@@ -164,6 +164,6 @@ func pushCmd(c *cli.Context) error {
 		return err
 	}
 
-	//return runtime.PushImage(srcName, destName, options)
-	return newImage.PushImage(getContext(), destName, manifestType, c.String("authfile"), c.String("signature-policy"), writer, c.Bool("compress"), so, &dockerRegistryOptions, forceSecure, nil)
+	//return runtime.PushImageToHeuristicDestination(srcName, destName, options)
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply cmd/podman/push.go]: Dropped.

**Comment by mtrmac** [comment]
> I'm a little discouraged that references have started to enter the API - the original intent was to try and insulate people from the large amount of potentially-confusing reference types as much as possible. But it doesn't seem like it can be avoided - attempting to do this with strings is like pulling teeth.

Yeah; using strings avoids the potentially-confusing reference _types_, but not the potentially-confusing reference _semantics_ – and it _adds_ the potentially-confusing reference _syntax_.

E.g. `cmd/podman/save.go` now calls more APIs, but embeds exactly the knowledge about the _subject_ of OCI archives in
```go
		dst = libpod.OCIArchive + ":" + output
…
 		// prepend = …
 		dest = fmt.Sprintf("%s:%s%s", dst, prepend, source)
```
vs.
```go
		destImageName := …
 		destRef, err = ociarchive.NewReference(output, destImageName) // destImageName may be ""
```
; but it now does not need to know about the `:` separator, and worry about misinterpreting extra `:` characters (after `docker/archive` is fixed, it should be possible to drop the `validateFileName` call which embeds another specific knowledge about the transports).

---

Or do you perhaps mean that you want to keep `libpod` as a reusable library, and not to expose its reliance on `c/image` to users? I could add wrappers for the API (`libpod.ImageReference` and `libpod.NewOCIArchiveReference`, or whatever you want); as long as we get rid of the string concatenation and parsing.

**Comment by mheon** [comment]
@mtrmac The original intent was to try and insulate users from c/image, but I think we're tightly coupled enough that it's not worth wrapping c/image API types.

The original goal was to create a situation where you would pass in an easy string, ideally straight from the user, and all the work of translating that into a reasonable input to c/image in the library. However, like you said, we just ended up with equally confusing string semantics, so it really did not help any.

**Comment by mtrmac** [comment]
> The original goal was to create a situation where you would pass in an easy string, ideally straight from the user, and all the work of translating that into a reasonable input to c/image in the library.

That still exists, e.g. as (using the proposed naming) `PushImageToHeuristicDestination`; but code that does _not_ take an input straight from the user and knows more about what the input is should not throw that knowledge away and subject the known input to heuristics.

**Comment by mheon** [comment]
Code LGTM

File: libpod/runtime_img.go [hunk 3]
```diff
@@ -21,27 +16,6 @@ import (
 
 // Runtime API
 
-var (
-	// DockerArchive is the transport we prepend to an image name
```
Comment (TomSweeneyRedHat) [inline libpod/runtime_img.go]: :+1: on pulling this from containers/image.  Very nice change.

**Comment by rhatdan** [comment]
@mtrmac Please remove the WIP and we can merge.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**